### PR TITLE
Feature Mautic integration with MS Dynamics CRM

### PR DESCRIPTION
**Pull request by virlatinus**
| Q  | A
| --- | ---
| Bug fix? | N
| New feature? | Y
| Related user documentation PR URL | https://mautic.org/docs/en/plugins/dynamics_crm.html
| Related developer documentation PR URL | N/A
| Issues addressed (#s or URLs) | https://github.com/mautic/mautic/issues/331
| BC breaks? | N/A
| Deprecations? | N/A

#### Description:
This PR adds a new integration with Microsoft Dynamics CRM.

#### Steps to test this PR:
1. Apply this PR
2. Clear the cache
3. Go to plugins and click on the install button
4. Configure the plugin with MSD URL and credentials.

#### Set Up Dynamics 365:
1. Go to the [Dynamics 365 Trial website](https://www.microsoft.com/en-us/dynamics/free-crm-trial.aspx)

image
image
image
image
image

#### Set Up Azure
1. Go to the [Azure Portal](https://portal.azure.com)
2. Log in with your onmicrosoft.com account

image

3. Go to Azure Active Directory 

image

4. Add a new Application Registration

image

5. Fill in the CRM Application information

image

6. Click on Create
7. Click on the Application you just created

image

8. You will use the Application ID when configuring the plugin in Mautic

image

9. Add a new Key. Use any name, click on save and copy the value. You will use it as the plugin secret in Mautic.

image
image

10. Configure the reply URLs using the callbacks from the plugin settings in Mautic. Click Save

image

11. Configure the Required Permissions. Click on Add 

image

12. Add Dynamics CRM Online Api Access. Click Select

image

13. Enable Dynamics CRM access for the users. Click Select and then click Done

image

15. Activate the permissions by clicking "Grant Permissions". Click Yes

image

16. Go back to Mautic
17. Authorize the plugin

image

18. Use your onmicrosoft.com account to authenticate:

image

19. The plugin is ready. You can test using "Push to Integration" form and campaign actions.
20. You can also test by executing the command: `php app/console mautic:integration:fetchleads -i Dynamics`

### CODE REVIEW

**Comment by alanhartless** [comment]
@virlatinus Let's add this to the existing MauticCrmBundle plugin rather than create a new whole new plugin for it. This keeps all our core Mautic supported CRMs together.

File: composer.json [hunk 1]
```diff
@@ -106,7 +106,11 @@
         "sonata-project/exporter": "^1.7",
         "lightsaml/sp-bundle": "~1.0.3",
         "symfony/expression-language": "~2.8",
-        "egeloen/ordered-form-bundle": "^3.0"
+        "egeloen/ordered-form-bundle": "^3.0",
+
+        "monolog/monolog": "^1.22",
+        "nesbot/carbon": "^1.22",
+        "illuminate/support": "^5.4"
```
Comment (escopecz) [inline composer.json]: One of the new dependencies is having problem with PHP 5.6: https://travis-ci.org/mautic/mautic/jobs/246235552#L181

Comment (virlatinus) [inline composer.json]: I removed the conflicting dependencies. They were not needed.

**Comment by dongilbert** [comment]
It seems that duplicates are created in the CRM when running it multiple times. Do we not match against email address? Also, when running the cli commands, the entity manager closes

**Comment by javiergas** [comment]
I updated to 2.9, configured the integration with my Dynamics 365 online and a campaign to Push contacts on read email. Now i'm getting this:
image

Found this on the logs:
[2017-07-18 17:58:40] mautic.NOTICE: PHP Notice - Undefined index: contacts - in file /opt/bitnami/apps/mautic/htdocs/app/bundles/PluginBundle/Integration/AbstractIntegration.php - at line 1795 [] []
[2017-07-18 17:58:40] mautic.WARNING: PHP Warning - Invalid argument supplied for foreach() - in file /opt/bitnami/apps/mautic/htdocs/app/bundles/PluginBundle/Integration/AbstractIntegration.php - at line 1801 [] []

By excuting the command php app/console mautic:integration:fetchleads -i Dynamics
it Works fine:

app/console mautic:integration:fetchleads -i Dynamics
Command to fetch contacts from integration
Fetch contacts command is starting

0 contacts were updated and 0 contacts were created

Updating/creating leads from Mautic to Dynamics
About 0 to update and about 1 to create/update
This could take some time. Please wait until the process is completed
 1/1 [============================] 100%
Number of contacts pushed: 0 updated, 1 created, 0 had errors and 0 were ignored (likely duplicates or didn't match field criteria)

**Comment by virlatinus** [comment]
Hi @javiergas . Please report it as an issue so we can take a look at it. https://github.com/mautic/mautic/issues
(Did you map your fields correctly? Are you synchronizing both contacts and companies? )

**Comment by javiergas** [comment]
Thank you very much. Here is the issue:
https://github.com/mautic/mautic/issues/4467

**Comment by javiergas** [comment]
Hi @virlatinus,
I updated to 2.9.2 and the issue is still there.
I'll be grad to help you debug this.

**Pull request merged**